js/runner.go

```diff
@@ -356,65 +353,46 @@ func (r *Runner) timeoutErrorDuration(stage string) time.Duration {
 type VU struct {
 	BundleInstance
 
-	Runner    *Runner
-	Transport *http.Transport
-	Dialer    *netext.Dialer
-	CookieJar *cookiejar.Jar
-	TLSConfig *tls.Config
-	ID        int64
-	Iteration int64
+	Runner     *Runner
+	RunContext *context.Context
+	Transport  *http.Transport
+	Dialer     *netext.Dialer
+	CookieJar  *cookiejar.Jar
+	TLSConfig  *tls.Config
+	ID         int64
+	Iteration  int64
 
 	Console *console
 	BPool   *bpool.BufferPool
 
 	Samples chan<- stats.SampleContainer
 
 	setupData goja.Value
-
-	// A VU will track the last context it was called with for cancellation.
-	// Note that interruptTrackedCtx is the context that is currently being tracked, while
-	// interruptCancel cancels an unrelated context that terminates the tracking goroutine
-	// without triggering an interrupt (for if the context changes).
-	// There are cleaner ways of handling the interruption problem, but this is a hot path that
-	// needs to be called thousands of times per second, which rules out anything that spawns a
-	// goroutine per call.
-	interruptTrackedCtx context.Context
-	interruptCancel     context.CancelFunc
-
-	m *sync.Mutex
 }
 
-// Verify that VU implements lib.VU
-var _ lib.VU = &VU{}
+// Verify that interfaces are implemented
+var _ lib.ActiveVU = &VU{}
+var _ lib.InitializedVU = &VU{}
 
-func (u *VU) Reconfigure(id int64) error {
-	u.ID = id
-	u.Iteration = 0
-	u.Runtime.Set("__VU", u.ID)
-	return nil
-}
+// Activate the VU so it will be able to run code
+func (u *VU) Activate(params *lib.VUActivationParams) lib.ActiveVU {
+	u.Runtime.ClearInterrupt()
+	u.RunContext = &params.RunContext
+	// u.Env = params.Env
 
-func (u *VU) RunOnce(ctx context.Context) error {
-	u.m.Lock()
-	defer u.m.Unlock()
-	// Track the context and interrupt JS execution if it's cancelled.
-	if u.interruptTrackedCtx != ctx {
-		interCtx, interCancel := context.WithCancel(context.Background())
-		if u.interruptCancel != nil {
-			u.interruptCancel()
+	go func() {
+		<-params.RunContext.Done()
+		u.Runtime.Interrupt(errInterrupt)
```

Comment:
Hmm I'm wondering if we shouldn't have more synchronization here - you're not waiting for this interruption to actually interrupt any potentially concurrently executing `RunOnce()` call. So, theoretically, I think the VU can currently be reused before it has finished executing the previous iteration from some other executor?

Reply:
Good point. Something like a `RunMutex` maybe?

Reply:
hmm not sure if a `Mutex` would be enough here :confused: this goroutine needs to wait for any execution currently running to finish after the `Interrupt()`, but there might not be any execution at the time... maybe a mutex would be enough, I'm not certain

Reply:
Partially fixed by e1fcc20a?

Yeah, a mutex wouldn't have worked in this case. I debated whether to use a `WaitGroup`, but that was messier than this `Done chan struct{}` approach.

Though judging from the current CI results, `<-vusDone` is executed more times than needed, leading to hanging somewhere. I'll look into it on Monday.